From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dongwon Kim <dongwon.kim@intel.com>
Date: Fri, 27 Aug 2021 13:06:52 -0700
Subject: [PATCH] ui/gtk-egl: dmabuf scanout only when scanout_mode is on in
 gd_egl_draw

dmabuf buf related code path (checkes on flag and mutex protection) only
if scanout_mode is set as it's not relevant to the normal draw path.

Signed-off-by: Dongwon Kim <dongwon.kim@intel.com>
---
 ui/gtk-egl.c | 44 ++++++++++++++++++++------------------------
 1 file changed, 20 insertions(+), 24 deletions(-)

diff --git a/ui/gtk-egl.c b/ui/gtk-egl.c
index 94e8649747..298b7c8f4e 100644
--- a/ui/gtk-egl.c
+++ b/ui/gtk-egl.c
@@ -74,25 +74,31 @@ void gd_egl_draw(VirtualConsole *vc)
     ww = gdk_window_get_width(window);
     wh = gdk_window_get_height(window);
 
-    if (dmabuf) {
-       qemu_mutex_lock(&dmabuf->mutex);
-       if (!dmabuf->draw_submitted) {
-           qemu_mutex_unlock(&dmabuf->mutex);
-           return;
-       } else {
-          dmabuf->draw_submitted = false;
-       }
-    }
     if (vc->gfx.scanout_mode) {
+        if (dmabuf) {
+            if (!dmabuf->draw_submitted) {
+                return;
+            } else {
+                dmabuf->draw_submitted = false;
+            }
+        }
         gd_egl_scanout_flush(&vc->gfx.dcl, 0, 0, vc->gfx.w, vc->gfx.h);
 
         vc->gfx.scale_x = (double)ww / vc->gfx.w;
         vc->gfx.scale_y = (double)wh / vc->gfx.h;
+
+        glFlush();
+        if (dmabuf) {
+            egl_dmabuf_create_fence(dmabuf);
+            if (dmabuf->fence_fd > 0) {
+                qemu_set_fd_handler(dmabuf->fence_fd, gd_hw_gl_flushed, NULL, vc);
+                return;
+            }
+            graphic_hw_gl_block(vc->gfx.dcl.con, false);
+        }
+        graphic_hw_gl_flushed(vc->gfx.dcl.con);
     } else {
         if (!vc->gfx.ds) {
-            if (dmabuf) {
-                qemu_mutex_unlock(&dmabuf->mutex);
-            }
             return;
         }
         eglMakeCurrent(qemu_egl_display, vc->gfx.esurface,
@@ -106,20 +112,10 @@ void gd_egl_draw(VirtualConsole *vc)
         gd_gl_count_frame(&vc->gfx.dcl, false);
         vc->gfx.scale_x = (double)ww / surface_width(vc->gfx.ds);
         vc->gfx.scale_y = (double)wh / surface_height(vc->gfx.ds);
-    }
 
-    glFlush();
-    if (dmabuf) {
-        egl_dmabuf_create_fence(dmabuf);
-        if (dmabuf->fence_fd > 0) {
-            qemu_set_fd_handler(dmabuf->fence_fd, gd_hw_gl_flushed, NULL, vc);
-	    qemu_mutex_unlock(&dmabuf->mutex);
-            return;
-        }
-        graphic_hw_gl_block(vc->gfx.dcl.con, false);
-	qemu_mutex_unlock(&dmabuf->mutex);
+        glFlush();
+        graphic_hw_gl_flushed(vc->gfx.dcl.con);
     }
-    graphic_hw_gl_flushed(vc->gfx.dcl.con);
 }
 
 void gd_egl_update(DisplayChangeListener *dcl,
-- 
2.17.1

